File: bindings/dart/lib/src/proxy.dart
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) 2026 Jonathan D.A. Jewell (hyperpolymath) <j.d.a.jewell@open.ac.uk>
//
// Proxy protocol types for proven-servers.

/// ProxyMode matching the Idris2 ABI tags.
enum ProxyMode {
  forward(0),
  reverse(1);

  const ProxyMode(this.tag);
  final int tag;

  static ProxyMode? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// HopByHopHeader matching the Idris2 ABI tags.
enum HopByHopHeader {
  connection(0),
  keepAlive(1),
  proxyAuth(2),
  proxyAuthz(3),
  te(4),
  trailers(5),
  transferEncoding(6),
  upgrade(7);

  const HopByHopHeader(this.tag);
  final int tag;

  static HopByHopHeader? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// CacheDirective matching the Idris2 ABI tags.
enum CacheDirective {
  noCache(0),
  noStore(1),
  maxAge(2),
  public(3),
  private(4),
  mustRevalidate(5);

  const CacheDirective(this.tag);
  final int tag;

  static CacheDirective? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

/// ProxyError matching the Idris2 ABI tags.
enum ProxyError {
  badGateway(0),
  gatewayTimeout(1),
  upstreamRefused(2),
  upstreamTls(3);

  const ProxyError(this.tag);
  final int tag;

  static ProxyError? fromTag(int tag) {
    for (final v in values) {
      if (v.tag == tag) return v;
    }
    return null;
  }
}

File: bindings/javascript/src/proxy.js
// SPDX-License-Identifier: MPL-2.0
// Copyright (c) Jonathan D.A. Jewell <j.d.a.jewell@open.ac.uk>
//
// Proxy protocol types for proven-servers.

/** ProxyMode matching the Idris2 ABI tags. */
export const ProxyMode = Object.freeze({
  FORWARD: 0,
  REVERSE: 1,
});

/** HopByHopHeader matching the Idris2 ABI tags. */
export const HopByHopHeader = Object.freeze({
  CONNECTION: 0,
  KEEP_ALIVE: 1,
  PROXY_AUTH: 2,
  PROXY_AUTHZ: 3,
  TE: 4,
  TRAILERS: 5,
  TRANSFER_ENCODING: 6,
  UPGRADE: 7,
});

/** CacheDirective matching the Idris2 ABI tags. */
export const CacheDirective = Object.freeze({
  NO_CACHE: 0,
  NO_STORE: 1,
  MAX_AGE: 2,
  PUBLIC: 3,
  PRIVATE: 4,
  MUST_REVALIDATE: 5,
});

/** ProxyError matching the Idris2 ABI tags. */
export const ProxyError = Object.freeze({
  BAD_GATEWAY: 0,
  GATEWAY_TIMEOUT: 1,
  UPSTREAM_REFUSED: 2,
  UPSTREAM_TLS: 3,
});
